cmd/catalyst-node/catalyst-node.go

```diff
@@ -348,3 +359,66 @@ func main() {
 
 	os.Exit(exitCode)
 }
+
+func startCatalystWebServer(httpAddr string) {
+	http.Handle("/hls/", redirectHlsHandler())
+	glog.Infof("HTTP server listening on %s", httpAddr)
+	http.ListenAndServe(httpAddr, nil)
+}
+
+var getClosestNode = queryMistForClosestNode
+
+func redirectHlsHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		playbackID, isValid := parsePlaybackID(r.URL.Path)
+		if !isValid {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+
+		nodeAddr, err := getClosestNode(playbackID)
+		if err != nil {
+			glog.Error(err)
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+
+		redirectUrl := fmt.Sprintf("%s://%s/hls/%s/index.m3u8", protocol(r), nodeAddr, playbackID)
+		http.Redirect(w, r, redirectUrl, http.StatusFound)
+	})
+}
+
+func parsePlaybackID(path string) (string, bool) {
+	r := regexp.MustCompile("^/hls/([a-zA-Z0-9_+]+)/index.m3u8$")
+	m := r.FindStringSubmatch(path)
+	if len(m) < 2 {
+		return "", false
+	}
+	return m[1], true
+}
+
+func protocol(r *http.Request) string {
+	if r.Header.Get("X-Forwarded-Proto") == "https" {
+		return "https"
+	}
+	return "http"
+}
+
+func queryMistForClosestNode(playbackID string) (string, error) {
+	url := fmt.Sprintf("http://localhost:%s/%s", mistUtilLoadPort, playbackID)
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", errors.New(fmt.Sprintf("GET request '%s' failed with http status code %d", url, resp.StatusCode))
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", errors.New(fmt.Sprintf("GET request '%s' failed while reading response body", url))
+	}
+
```

Comment:
If the stream isn't available — or (less likely) there are no servers with the resources to handle the playback, MistUtilLoad will return the string "FULL". Need to handle that error case, seems potentially appropriate for this function?

Reply:
Updated in 219d1f07fb8e7d35f8069c7af847ea8bb138de2e